docs/ha-considerations.md

```diff
@@ -0,0 +1,332 @@
+# High Availability Considerations
+
+## Overview
+
+When setting up a production cluster, high availability (the cluster's ability to remain operational even if some control plane or worker nodes fail) is usually a requirement. For worker nodes, assuming that there are enough of them, this is part of the very cluster functionality. However redundancy of control plane nodes and `etcd` instances needs to be catered for when planning and setting up a cluster.
+
+`kubeadm` supports setting up of multi control plane and multi `etcd` clusters (see [Creating Highly Available clusters with kubeadm](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/high-availability/) for a step-by-step guide). Still there are some aspects to consider and set up which are not part of Kubernetes itself and hence not covered in the project documentation. This document provides some additional information and examples useful when planning and bootstrapping HA clusters with `kubeadm`.
+
+<!-- TOC -->
+
+- [High Availability Considerations](#high-availability-considerations)
+    - [Overview](#overview)
+    - [Options for Software Load Balancing](#options-for-software-load-balancing)
+    - [keepalived and haproxy](#keepalived-and-haproxy)
+    - [kube-vip](#kube-vip)
+    - [Bootstrap the cluster](#bootstrap-the-cluster)
+
+## Options for Software Load Balancing
+
+When setting up a cluster with more than one control plane, higher availability can be achieved by putting the API Server instances behind a load balancer and using the `--control-plane-endpoint` option when running `kubeadm init` for the new cluster to use it.
+
+Of course, the load balancer itself should be highly available, too. This is usually achieved by adding redundancy to the load balancer. In order to do so, a cluster of hosts managing a virtual IP is set up with each host running an instance of the load balancer, so that always the load balancer on the host currently holding the vIP will be used while the others are on standby.
+
+In some environments, like in data centers with dedicated load balancing components (provided e.g. by some cloud-providers), this functionality may already be available. If it is not, user-managed load balancing can be used. In that case some preparation is necessary before bootstrapping a cluster.
+
+Since this is infrastructure Kubernetes runs on and hence is not part of Kubernetes or `kubeadm`, this must be taken care of separately. In the following sections we give examples that have been working for some people while of course there are potentially dozens of other possible configurations.
```

Comment:
```suggestion
Since this is not part of Kubernetes nor in the scope of `kubeadm`, this must be taken care of separately. In the following sections, we give examples that have been working for some people while of course there are potentially dozens of other possible configurations.
```

Reply:
(y)